File: solutions/cpp/rotational-cipher/1/rotational_cipher.cpp
#include "rotational_cipher.h"
#include <string>
#include <cctype>

namespace rotational_cipher {

std::string rotate(std::string text, int shift_key) {
    // Normalize the shift key to handle values > 26 or < 0
    shift_key = shift_key % 26;
    if (shift_key < 0) {
        shift_key += 26;  // Handle negative shifts
    }
    
    std::string result;
    result.reserve(text.length());  // Pre-allocate memory for efficiency
    
    for (char ch : text) {
        if (std::isalpha(ch)) {
            char base = std::islower(ch) ? 'a' : 'A';
            // Apply shift with modulo 26 to wrap around alphabet
            char shifted = static_cast<char>((ch - base + shift_key) % 26 + base);
            result.push_back(shifted);
        } else {
            // Non-alphabetic characters remain unchanged
            result.push_back(ch);
        }
    }
    
    return result;
}

}  // namespace rotational_cipher

File: solutions/python/rotational-cipher/1/rotational_cipher.py
def rotate(text, key):
    result = []
    key = key % 26  # Normalize key to 0-25
    
    for char in text:
        if char.isalpha():
            # Determine the base (A for uppercase, a for lowercase)
            base = ord('A') if char.isupper() else ord('a')
            # Shift the character and wrap around using modulo 26
            shifted = (ord(char) - base + key) % 26
            result.append(chr(base + shifted))
        else:
            # Keep non-alphabet characters unchanged
            result.append(char)
    
    return ''.join(result)